mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -265,101 +293,145 @@ impl Watcher {
-            .extract_secret(&self.data.secret_hash[..], &tx_ident.tx_hex.0)
+            .extract_secret(&watcher_ctx.data.secret_hash[..], &tx_ident.tx_hex.0)
         {
             Ok(bytes) => H256Json::from(bytes.as_slice()),
             Err(e) => {
-                return Ok((Some(WatcherCommand::Finish), vec![
-                    WatcherEvent::TakerPaymentWaitForSpendFailed(ERRL!("{}", e).into()),
-                ]));
+                return Self::change_state(Stopped::from_reason(StopReason::TakerPaymentWaitForSpendFailed(
+                    ERRL!("{}", e).into(),
+                )))
             },
         };
 
-        Ok((Some(WatcherCommand::SpendMakerPayment), vec![
-            WatcherEvent::TakerPaymentSpent(TakerPaymentSpentData {
-                transaction: tx_ident,
-                secret,
-            }),
-        ]))
+        Self::change_state(SpendMakerPayment::new(secret))
     }
+}
 
-    async fn spend_maker_payment(&self) -> Result<(Option<WatcherCommand>, Vec<WatcherEvent>), String> {
-        let spend_fut = self.maker_coin.send_taker_spends_maker_payment_preimage(
-            &self.data.taker_spends_maker_payment_preimage,
-            &self.r().secret.0.clone(),
+#[async_trait]
+impl State for SpendMakerPayment {
+    type Ctx = WatcherContext;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, watcher_ctx: &mut WatcherContext) -> StateResult<Self::Ctx, Self::Result> {
+        let spend_fut = watcher_ctx.maker_coin.send_taker_spends_maker_payment_preimage(
+            &watcher_ctx.data.taker_spends_maker_payment_preimage,
+            &self.secret.0,
         );
 
         let transaction = match spend_fut.compat().await {
             Ok(t) => t,
             Err(err) => {
                 if let Some(tx) = err.get_tx() {
-                    broadcast_p2p_tx_msg(&self.ctx, tx_helper_topic(self.maker_coin.ticker()), &tx, &None);
+                    broadcast_p2p_tx_msg(
+                        &watcher_ctx.ctx,
+                        tx_helper_topic(watcher_ctx.maker_coin.ticker()),
+                        &tx,
+                        &None,
+                    );
                 };
-
-                return Ok((Some(WatcherCommand::Finish), vec![
-                    WatcherEvent::MakerPaymentSpendFailed(ERRL!("{}", err.get_plain_text_format()).into()),
-                ]));
+                return Self::change_state(Stopped::from_reason(StopReason::MakerPaymentSpendFailed(
+                    ERRL!("{}", err.get_plain_text_format()).into(),
+                )));
             },
         };
 
         broadcast_p2p_tx_msg(
-            &self.ctx,
-            tx_helper_topic(self.maker_coin.ticker()),
+            &watcher_ctx.ctx,
+            tx_helper_topic(watcher_ctx.maker_coin.ticker()),
             &transaction,
             &None,
         );
 
         let tx_hash = transaction.tx_hash();
         info!("Maker payment spend tx {:02x}", tx_hash);
```

Comment:
To distinguish logging outputs, it's worth changing it to something like `Sent maker payment spend tx {:02x} as wathcher`.

Reply:
@caglaryucekaya Could you please change the message for one more line?
https://github.com/KomodoPlatform/atomicDEX-API/blob/watchtower/mm2src/mm2_main/src/lp_swap/swap_watcher.rs#L295